Detect linux version at runtime (for worked nested epoll)

azat opened the pull request:
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Detect linux version at runtime (for worked nested epoll, that is required for `async_socket_for_remote`/`use_hedged_requests`, otherwise remote queries may stuck)

Detailed description / Documentation draft:
The #22109 adds the check but at compilation time, which is pointless,
move the check into runtime.

Remember nested epoll is required for
async_socket_for_remote/use_hedged_requests, otherwise remote queries
may stuck.

azat committed:
Detect linux version at runtime (for worked nested epoll)

The #22109 adds the check but at compilation time, which is pointless,
move the check into runtime.

Remember nested epoll is required for
async_socket_for_remote/use_hedged_requests, otherwise remote queries
may stuck.

azat commented on src/Common/VersionNumber.cpp:
```diff
@@ -0,0 +1,61 @@
+VersionNumber::VersionNumber(const std::vector<long> & vec)
+{
+    if (vec.size() > SIZE)
+        throw Exception(ErrorCodes::BAD_VERSION, "Too much components ({})", vec.size());
```
But `LOGICAL_ERROR` cannot be covered with unit tests, since it calls `abort()`

kitaisreal replied:
You can probably use ASSERT_DEATH in gtest to test this.

kitaisreal commented on src/Common/VersionNumber.h:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    /// @param strict - throws if number of components > 3
+    static VersionNumber fromString(std::string version, bool strict);
```
Why static constructor ? Probably better to just create ordinary constructor.

azat replied:
Logically parsing from string sounds different to me, but I don't have strong opinion on this, will rewrite to constructor.

azat replied:
Moved fromString into another constructor.

azat committed:
Use LOGICAL_ERROR instead of introducing BAD_VERSION

But note, that tests with EXPECT_EXIT() does not work with --gtest_filter:

a) only this test is included into unit_tests_dbms

   $ src/unit_tests_dbms --gtest_filter=VersionNumber*
   Running main() from ../contrib/googletest/googletest/src/gtest_main.cc
   Note: Google Test filter = VersionNumber*
   [==========] Running 2 tests from 1 test suite.
   [----------] Global test environment set-up.
   [----------] 2 tests from VersionNumber
   [ RUN      ] VersionNumber.VersionNumber
   [       OK ] VersionNumber.VersionNumber (0 ms)
   [ RUN      ] VersionNumber.fromString
   Segmentation fault (core dumped)

   (gdb) bt
   0  0x00007fffda323d22 in raise () from /usr/lib/libc.so.6
   1  0x00007fffda30d862 in abort () from /usr/lib/libc.so.6
   2  0x00007fffdbe686f5 in DB::handle_error_code (msg=..., code=49, remote=false, trace=...) at Exception.cpp:49
   3  0x00007fffdbe68846 in DB::Exception::Exception (this=0x7fffd5c5a240, msg=..., code=49, remote_=false) at Exception.cpp:60
   4  0x00007fffe26b2cb3 in DB::Exception::Exception<unsigned long> (this=0x7fffd5c5a240, code=49, fmt=..., args=@0x7fffffffc7e8: 4) at Exception.h:40
   5  0x00007fffdbf4d201 in DB::VersionNumber::VersionNumber (this=0x7fffffffcc20, vec=...) at VersionNumber.cpp:17
   6  0x00007fffdbf4d650 in DB::VersionNumber::fromString (version=..., strict=true) at VersionNumber.cpp:57
   7  0x00005555555db53d in VersionNumber_fromString_Test::TestBody (this=0x7fffd5c12330) at gtest_version_number.cpp:24

b) regular build

   $ src/unit_tests_dbms --gtest_filter=VersionNumber*
   Running main() from ../contrib/googletest/googletest/src/gtest_main.cc
   Note: Google Test filter = VersionNumber*
   [==========] Running 2 tests from 1 test suite.
   [----------] Global test environment set-up.
   [----------] 2 tests from VersionNumber
   [ RUN      ] VersionNumber.VersionNumber
   [       OK ] VersionNumber.VersionNumber (0 ms)
   [ RUN      ] VersionNumber.fromString
   [       OK ] VersionNumber.fromString (495 ms)
   [----------] 2 tests from VersionNumber (495 ms total)

   [----------] Global test environment tear-down
   Segmentation fault (core dumped)

   (gdb) bt
   0  testing::TestInfo::should_run (this=0xe0) at gtest.h:760
   1  0x00007ffff7f6edd5 in testing::TestSuite::ShouldRunTest (test_info=0xe0) at gtest.h:1009
   2  0x00007ffff7f6ebb2 in testing::internal::CountIf<std::__1::vector<testing::TestInfo*, std::__1::allocator<testing::TestInfo*> >, bool (*)(testing::TestInfo const*)> (c=..., predicate=0x7ffff7f6edc0 <testing::TestSuite::ShouldRunTest(testing::TestInfo const*)>)
      at gtest-internal-inl.h:294
   3  0x00007ffff7f41ae5 in testing::TestSuite::test_to_run_count (this=0x7fffd5028500) at gtest.cc:2919
   4  0x00007ffff7f41719 in testing::internal::SumOverTestSuiteList (case_list=..., method=(int (testing::TestSuite::*)(const testing::TestSuite * const)) 0x7ffff7f41ac0 <testing::TestSuite::test_to_run_count() const>) at gtest.cc:377
   5  0x00007ffff7f41ab9 in testing::internal::UnitTestImpl::test_to_run_count (this=0x7fffd5d5c000) at gtest.cc:1012
   6  0x00007ffff7f4b59d in testing::UnitTest::test_to_run_count (this=0x7ffff7fbcb80 <testing::UnitTest::GetInstance()::instance>) at gtest.cc:5059
   7  0x00007ffff7f4c4ed in testing::internal::PrettyUnitTestResultPrinter::OnTestIterationEnd (this=0x7fffd5ca4a68, unit_test=...) at gtest.cc:3589
   8  0x00007ffff7f4daf9 in testing::internal::TestEventRepeater::OnTestIterationEnd (this=0x7fffd5c83770, unit_test=..., iteration=0) at gtest.cc:3852
   9  0x00007ffff7f57570 in testing::internal::UnitTestImpl::RunAllTests (this=0x7fffd5d5c000) at gtest.cc:5737
   10 0x00007ffff7fa3a84 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x7fffd5d5c000,
      method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x7ffff7f56f20 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x7ffff7f11d9b "auxiliary test code (environments or event listeners)") at gtest.cc:2589
   11 0x00007ffff7f71788 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x7fffd5d5c000,
      method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x7ffff7f56f20 <testing::internal::UnitTestImpl::RunAllTests()>, location=0x7ffff7f11d9b "auxiliary test code (environments or event listeners)") at gtest.cc:2625
   12 0x00007ffff7f56ea3 in testing::UnitTest::Run (this=0x7ffff7fbcb80 <testing::UnitTest::GetInstance()::instance>) at gtest.cc:5291

kitaisreal commented on src/Common/VersionNumber.h:
```diff
@@ -0,0 +1,52 @@
+/// Supports only "major.minor.patch", other components are ignored.
+struct VersionNumber
+{
+    using VersionTuple = std::tuple<long, long, long>;
```
Move to private.

kitaisreal commented on src/Core/SettingsQuirks.cpp:
```diff
@@ -14,19 +15,22 @@
+        Poco::Environment::osVersion(), /* strict= */ false);
+
+    /// the check is correct since there will be no more 5.5.x releases.
+    if (linux_version >= DB::VersionNumber{5, 5, 0} && linux_version < DB::VersionNumber{5, 6, 13})
```
@azat did you check that it will work as expected ? And Poco::Environment::osVersion().
Try to use this in reneameat2.

tavplubix replied:
Btw, `uname` and `supportsRenameat2Impl()` do not always work as expected (so `Poco::Environment::osVersion()` and `DB::VersionNumber` will not too, because it works similar way). The problem is that WSL and Docker on Mac pretend to be an ordinary Linux kernel, but actually they are not and behave differently. For example see 
https://github.com/ClickHouse/ClickHouse/blob/549b5f8a0cd037bc9032d976921d64500bfe52e6/src/Common/renameat2.cpp#L66-L75

azat replied:
> The problem is that WSL and Docker on Mac pretend to be an ordinary Linux kernel, but actually they are not and behave differently

Right, but there is not a lot, that can be done in this case.

> did you check that it will work as expected ?

Yes.

> Try to use this in reneameat2.

Ok

azat replied:
>Try to use this in reneameat2.

Done

alexey-milovidov commented on src/Common/VersionNumber.h:
```diff
@@ -0,0 +1,52 @@
+/// Simple numeric version representation.
+///
+/// Supports only "major.minor.patch", other components are ignored.
+struct VersionNumber
```
I don't like this struct, let's move it completely to SettingsQuirks.cpp

azat replied:
Do you dislike something specific?
Initially I placed it in SettingsQuirks, but later decided that it is common enough and that I want to cover it with unit test.
Plus there is another such place - renameat2 detection, that may use the same code.

alexey-milovidov commented on src/Core/SettingsQuirks.cpp:
```diff
@@ -14,19 +15,22 @@
+
+    /// the check is correct since there will be no more 5.5.x releases.
+    if (linux_version >= DB::VersionNumber{5, 5, 0} && linux_version < DB::VersionNumber{5, 6, 13})
+    {
```
Let's make a plan on how soon we will remove this code completely.
Maybe 2022-01-01?

azat replied:
Well, if something that does not in control of clickhouse has an issue it should not be removed until it will past enough time (1-2 years).
Consider linux kernel, it has lot's of quirks for hardware.

And also note, that this is pretty nasty issue (I spend quite a lot of time to understand that the issue is in the kernel not in the async_socket_for_remote or/and use_hedged_requests), so that said that to me, this quirk brings more benefits than break "code clearness".

alexey-milovidov commented on src/Core/SettingsQuirks.cpp:
```diff
@@ -14,19 +15,22 @@
+        Poco::Environment::osVersion(), /* strict= */ false);
+
+    /// the check is correct since there will be no more 5.5.x releases.
+    if (linux_version >= DB::VersionNumber{5, 5, 0} && linux_version < DB::VersionNumber{5, 6, 13})
```
Maybe simply refuse to start on these kernel versions instead of making quirks?

azat replied:
I don't think that this is a good idea, since the benefit of async_socket_for_remote is doubtful to me (it has lots of syscall overhead at least), and use_hedged_requests is significant only for big cluster setup.

alexey-milovidov commented on src/Common/VersionNumber.h:
```diff
@@ -0,0 +1,52 @@
+namespace DB
+{
+
+/// Simple numeric version representation.
```
Maybe just make a single function to compare versions?

azat committed:
Convert VersionNumber::fromString() to constructor

azat committed:
Use VersionNumber for detecting renameat2()

azat committed:
Set -Wno-covered-switch-default for unit_tests_dbms

Otherwise code that uses EXPECT_EXIT() fails with -Werror

v2: adjust only unit_tests_dbms, not googltest via
contrib/CMakeListst.txt itself.

alexey-milovidov commented:
Then let's make it support arbitrary number of components (simple std::vector)

alexey-milovidov commented:
The potential harm is that when people will see a file named VersionNumber they may start to think that we use the same class for ClickHouse version numbers.

alexey-milovidov commented:
So, to solve this issue, we can bury this class into some file where no one will look.

alexey-milovidov commented:
And if we do it, I don't care that it only supports three-component versions.

azat commented:
>Then let's make it support arbitrary number of components (simple std::vector)

Ok.
(I though about this initially, but decided to keep it simpler, by using tuple, i.e. compare is much simpler with it)

alexey-milovidov commented:
People also may mistakenly think that there is something good in this code, but actually it is not suitable for typical version numbers (e.g. OpenSSL version that have alphabetical components).

azat commented:
>The potential harm is that when people will see a file named VersionNumber they may start to think that we use the same class for ClickHouse version numbers.

That was one of my concerns, and this is the reason why it called VersionNumber not Version.

alexey-milovidov commented:
@tavplubix said that he has the required function somewhere...

alexey-milovidov commented:
I'm ok for this change if we bury the code as deep as possible...

azat committed:
More generic VersionNumber

azat committed:
Drop unnecessary qualified namespace

azat committed:
Fix clang-tidy warnings in VersionNumber (use Int64 over long)

azat commented:
>Then let's make it support arbitrary number of components (simple std::vector)

I made VersionNumber for arbitrary numbers, and now I think it is OK to keep it in `Common/VersoinNumber`. Objections?

alexey-milovidov commented:
No objections...
